### PR TITLE
Fix and improve unit conversion when writing NinJoTIFF

### LINKED ISSUE #1614: Saving AAPP-processed NOAA HRPT to NinJoTIFF fails with AttributeError
**Describe the bug**

When loading a NOAA AVHRR HRPT Level-1B file produced with AAPP and saving it to NinJoTIFF, pint raises an exception `AttributeError: 'NoneType' object has no attribute 'evaluate'`.

**To reproduce**

```python
from satpy import Scene
from satpy.utils import debug_on; debug_on()
sc = Scene(
        filenames=["/media/nas/x21308/scratch/hrpt_noaa19_20210324_0828_62464.l1b"],
        reader=["avhrr_l1b_aapp"])
sc.load(["1"])
ls = sc.resample("eurol")
ls.save_dataset("1",
        "/tmp/{platform_name}-{area.area_id}-{start_time:%Y%m%d_%H%M}-{end_time:%Y%m%d_%H%M}.tif",
        writer="ninjotiff",
        physic_value="ALBEDO",
        physic_unit="%",
        sat_id=7900014,
        chan_id=100015,
        data_cat="PORN",
        nbits=8,
        data_source="HRPT",
        ch_min_measurement_unit=0,
        ch_max_measurement_unit=125)
```

**Expected behaviour**

I expect that a NinJoTIFF file is written with the indicated information.

**Actual results**

```
[DEBUG: 2021-03-24 10:27:15 : satpy.readers.yaml_reader] Reading ('/home/gholl/checkouts/satpy/satpy/etc/readers/avhrr_l1b_aapp.yaml',)
[DEBUG: 2021-03-24 10:27:15 : satpy.readers.yaml_reader] Assigning to avhrr_l1b_aapp: ['/media/nas/x21308/scratch/hrpt_noaa19_20210324_0828_62464.l1b']
[DEBUG: 2021-03-24 10:27:15 : satpy.readers.aapp_l1b] Reading time 0:00:00.003219
[DEBUG: 2021-03-24 10:27:15 : satpy.composites.config_loader] Looking for composites config file avhrr-3.yaml
[DEBUG: 2021-03-24 10:27:15 : satpy.composites.config_loader] Looking for composites config file visir.yaml
[DEBUG: 2021-03-24 10:27:15 : satpy.readers.yaml_reader] No coordinates found for DataID(name='longitude', resolution=1050, modifiers=())
[DEBUG: 2021-03-24 10:27:15 : satpy.readers.yaml_reader] No coordinates found for DataID(name='latitude', resolution=1050, modifiers=())
[INFO: 2021-03-24 10:27:15 : satpy.readers.aapp_l1b] No valid operational coefficients, fall back to pre-launch
[DEBUG: 2021-03-24 10:27:15 : pyproj] PROJ_ERROR: proj_crs_get_sub_crs: Object is not a CompoundCRS
[DEBUG: 2021-03-24 10:27:15 : pyproj] PROJ_ERROR: proj_crs_get_sub_crs: Object is not a CompoundCRS
[DEBUG: 2021-03-24 10:27:15 : satpy.scene] Resampling DataID(name='1', wavelength=WavelengthRange(min=0.58, central=0.63, max=0.68, unit='µm'), resolution=1050, calibration=<calibration.reflectance>, modifiers=())
[INFO: 2021-03-24 10:27:15 : satpy.scene] Not reducing data before resampling.
[INFO: 2021-03-24 10:27:15 : satpy.resample] Using default KDTree resampler
/data/gholl/miniconda3/envs/py39/lib/python3.9/site-packages/pyresample/geometry.py:2118: DeprecationWarning: This function is deprecated. See: https://pyproj4.github.io/pyproj/stable/gotchas.html#upgrading-to-pyproj-2-from-pyproj-1
  hor_xyz = np.stack(transform(src, dst, x, mid_row_y, alt_x), axis=1)
/data/gholl/miniconda3/envs/py39/lib/python3.9/site-packages/pyresample/geometry.py:2119: DeprecationWarning: This function is deprecated. See: https://pyproj4.github.io/pyproj/stable/gotchas.html#upgrading-to-pyproj-2-from-pyproj-1
  vert_xyz = np.stack(transform(src, dst, mid_col_x, y, alt_y), axis=1)
[DEBUG: 2021-03-24 10:27:15 : satpy.resample] Check if ./resample_lut-b444378bd3f422170b51d20660d8af7fb545fd3f.npz exists
[DEBUG: 2021-03-24 10:27:15 : satpy.resample] Computing kd-tree parameters
/data/gholl/miniconda3/envs/py39/lib/python3.9/site-packages/pyresample/kd_tree.py:1025: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  input_coords = input_coords.astype(np.float)
[DEBUG: 2021-03-24 10:27:15 : pyproj] PROJ_ERROR: proj_crs_get_sub_crs: Object is not a CompoundCRS
[DEBUG: 2021-03-24 10:27:15 : pyproj] PROJ_ERROR: proj_crs_get_sub_crs: Object is not a CompoundCRS
/data/gholl/miniconda3/envs/py39/lib/python3.9/site-packages/pyresample/kd_tree.py:1047: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  radius=self.radius_of_influence, dtype=np.int,
[DEBUG: 2021-03-24 10:27:15 : satpy.resample] Resampling where-60d4d27451e64f80225b160f9967589e
[DEBUG: 2021-03-24 10:27:15 : satpy.writers] Reading ['/home/gholl/checkouts/satpy/satpy/etc/writers/ninjotiff.yaml']
/data/gholl/miniconda3/envs/py39/lib/python3.9/site-packages/pyninjotiff/tifffile.py:154: UserWarning: failed to import the optional _tifffile C extension module.
Loading of some compressed images will be slow.
Tifffile.c can be obtained at http://www.lfd.uci.edu/~gohlke/
  warnings.warn(
Traceback (most recent call last):
  File "/home/gholl/checkouts/protocode/mwe/ninjotiff-crash.py", line 8, in <module>
    ls.save_dataset("1",
  File "/home/gholl/checkouts/satpy/satpy/scene.py", line 993, in save_dataset
    return writer.save_dataset(self[dataset_id],
  File "/home/gholl/checkouts/satpy/satpy/writers/ninjotiff.py", line 182, in save_dataset
    dataset = convert_units(dataset, units, nunits)
  File "/home/gholl/checkouts/satpy/satpy/writers/ninjotiff.py", line 99, in convert_units
    in_unit = ureg.parse_expression(in_unit, False)
  File "/data/gholl/miniconda3/envs/py39/lib/python3.9/site-packages/pint/registry.py", line 1263, in parse_expression
    return build_eval_tree(gen).evaluate(
AttributeError: 'NoneType' object has no attribute 'evaluate'
```

**Environment info**

Python 3.9.2, satpy 0.26, pyresample 0.18.1, pyninjotiff 0.3.0, pint 0.17.

**Additional context**

I'm not sure if this is a problem in Satpy, pyninjotiff, or pint, but I think pyninjotiff is the prime suspect.

Saving AAPP-processed Metop AVHRR files to NinJoTIFF works fine.

Unfortunately the files in question are larger than 10 MB, so I cannot upload them here.

### PR DESCRIPTION
**Pull request by gerritholl**
Refactor, fix, improve the unit conversion that happens in the Satpy NinJoTIFF writer when writing a NinJoTIFF file.

 - [x] Closes #1614 
 - [x] Tests added 
 - [x] Fully documented

### CODE REVIEW

**Comment by gerritholl** [comment]
How is it possible that a test routine that mocks `convert_units` starts failing when I change the implementation of `convert_units`?

File: satpy/writers/ninjotiff.py [hunk 1]
```diff
@@ -90,23 +90,44 @@
...
-    attrs = dataset.attrs
-    dataset = data.to(dest_unit).magnitude
-    dataset.attrs = attrs
-    dataset.attrs["units"] = out_unit
-    return dataset
+    """Convert units of *dataset*.
+
+    Convert dataset units for the benefit of writing NinJoTIFF.  The main
+    background here is that NinJoTIFF would like brightness temperatures in °C,
+    but satellinge data files are in K.  For simplicity of implementation, this
```
Comment (mraspaud) [inline satpy/writers/ninjotiff.py:97]:
```suggestion
    but satellite data files are in K.  For simplicity of implementation, this
```

File: satpy/writers/ninjotiff.py [hunk 2]
```diff
@@ -90,23 +90,44 @@
...
+    function can only convert from °C to K.
```
Comment (mraspaud) [inline satpy/writers/ninjotiff.py:98]:
```suggestion
    function can only convert from K to °C.
```

File: satpy/writers/ninjotiff.py [hunk 3]
```diff
@@ -90,23 +90,44 @@
...
+
+    This function will convert input data from °C to K and write the new unit
```
Comment (mraspaud) [inline satpy/writers/ninjotiff.py:100]:
```suggestion
    This function will convert input data from K to °C and write the new unit
```

File: satpy/writers/ninjotiff.py [hunk 4]
```diff
@@ -90,23 +90,44 @@
...
+        new_dataset = dataset + 273.15
+        new_dataset.attrs["units"] = out_unit
+        return new_dataset
+
+    # Other cases not implemented.  Creating a quantity from a pint array
+    # doesn't work (TypeError: Quantity cannot wrap upcast type
+    # xarray.DataArray).  Working on the values may cause further bugs and
+    # dask-compatibility.  I don't know if anyone is using this function to
+    # convert between non-temperature units.
+    raise ValueError(
```
Comment (mraspaud) [inline satpy/writers/ninjotiff.py:128]:
```suggestion
    raise NotImplementedError(
```

File: satpy/writers/ninjotiff.py [hunk 5]
```diff
@@ -90,23 +90,44 @@
...
+    if in_unit.lower() in {"k", "kelvin"} and out_unit.lower() in {"c", "celsius"}:
+        new_dataset = dataset + 273.15
+        new_dataset.attrs["units"] = out_unit
+        return new_dataset
+
+    # Other cases not implemented.  Creating a quantity from a pint array
+    # doesn't work (TypeError: Quantity cannot wrap upcast type
+    # xarray.DataArray).  Working on the values may cause further bugs and
+    # dask-compatibility.  I don't know if anyone is using this function to
+    # convert between non-temperature units.
```
Comment (mraspaud) [inline satpy/writers/ninjotiff.py:127]: I don't think this comment is needed.

File: satpy/tests/writer_tests/test_ninjotiff.py [hunk 6]
```diff
@@ -82,3 +84,36 @@ def test_image(self, iwsi, save_dataset):
         assert(nt.save.mock_calls[0][2]['ch_min_measurement_unit']
                < nt.save.mock_calls[0][2]['ch_max_measurement_unit'])
         assert(ret == nt.save.return_value)
+
+    def test_convert_units(self):
```
Comment (mraspaud) [inline satpy/tests/writer_tests/test_ninjotiff.py:88]: Could you make three test functions here instead?

**Review by mraspaud — APPROVED** [review]
Thanks for fixing this, it was something that needed to be done for a while now :) A few comments.

File: satpy/writers/ninjotiff.py (hunk 1 above)
Comment (gerritholl) [reply satpy/writers/ninjotiff.py]: Fixed.

File: satpy/writers/ninjotiff.py (hunk 2 above)
Comment (gerritholl) [reply satpy/writers/ninjotiff.py]: Fixed.

File: satpy/writers/ninjotiff.py (hunk 3 above)
Comment (gerritholl) [reply satpy/writers/ninjotiff.py]: Fixed.

File: satpy/writers/ninjotiff.py (hunk 4 above)
Comment (gerritholl) [reply satpy/writers/ninjotiff.py]: Fixed.

File: satpy/writers/ninjotiff.py (hunk 5 above)
Comment (gerritholl) [reply satpy/writers/ninjotiff.py]: Shortened.

File: satpy/tests/writer_tests/test_ninjotiff.py (hunk 6 above)
Comment (gerritholl) [reply satpy/tests/writer_tests/test_ninjotiff.py]: Done.

**Comment by gerritholl** [comment]
The windows build fixed in setup, can we restart it?

**Comment by gerritholl** [comment]
Although this does close #1614, within trollflow2 I get a new failure now:

```
Traceback (most recent call last):
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/trollflow2/launcher.py", line 303, in process
 cwrk.pop('fun')(job, **cwrk)
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/trollflow2/plugins/__init__.py", line 260, in save_datasets
 obj = save_dataset(scns, fmat, fmat_config, renames)
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/trollflow2/plugins/__init__.py", line 219, in save_dataset
 obj = scns[fmat['area']].save_dataset(dsid,
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/satpy/scene.py", line 993, in save_dataset
 return writer.save_dataset(self[dataset_id],
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/satpy/writers/ninjotiff.py", line 202, in save_dataset
 return super(NinjoTIFFWriter, self).save_dataset(
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/satpy/writers/__init__.py", line 811, in save_dataset
 return self.save_image(img, filename=filename, compute=compute, fill_value=fill_value, **kwargs)
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/satpy/writers/ninjotiff.py", line 174, in save_image
 return nt.save(img, filename, data_is_scaled_01=True, compute=compute, **kwargs)
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/pyninjotiff/ninjotiff.py", line 541, in save
 area_def = img.data.area
 File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py39/lib/python3.9/site-packages/xarray/core/common.py", line 239, in __getattr__
 raise AttributeError(
AttributeError: 'DataArray' object has no attribute 'area'
```

I will try to construct a MCVE reproducing this problem and open a new issue (which, this time, I think, is in pyninjotiff).

**Comment by mraspaud** [comment]
Restarted the actions

**Pull request merged**